File: spark.scala
import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.types._
import org.apache.spark.sql.functions._
import org.apache.spark._
import java.sql.Timestamp

object OracleToHive {
  def main(args: Array[String]): Unit = {
    val spark = SparkSession.builder().appName("OracleToHive").getOrCreate()
    val jdbcUrl = "jdbc:oracle:thin:@hostname:port/service_name"
    val jdbcTable = "oracle_table"
    val jdbcUser = "username"
    val jdbcPassword = "password"
    val hiveTable = "hive_table"
    val hiveDatabase = "hive_database"
    val hiveFormat = "orc"
    val hivePartitionColumns = Seq("column1", "column2")
    val hivePartitionValues = Seq("value1", "value2")
    val hivePartitionLocation = "hdfs:///hive/warehouse/hive_database.db/hive_table"
    val streamingInterval = 10 //seconds

    //Get Oracle Table Schema
    val oracleTableSchema = spark.read.format("jdbc")
      .option("url", jdbcUrl)
      .option("dbtable", jdbcTable)
      .option("user", jdbcUser)
      .option("password", jdbcPassword)
      .option("driver", "oracle.jdbc.driver.OracleDriver")
      .load()
      .schema

    //Cast Oracle Column Data Types
    val castedSchema = oracleTableSchema.map(field => {
      field.dataType match {
        case StringType => StructField(field.name, StringType, field.nullable, field.metadata)
        case IntegerType => StructField(field.name, DecimalType(38, 0), field.nullable, field.metadata)
        case DateType => StructField(field.name, TimestampType, field.nullable, field.metadata)
        case _ => field
      }
    })

//Download Table in Hive with ORC Format
    val downloadedDF = spark.read.format("jdbc")
      .option("url", jdbcUrl)
      .option("dbtable", jdbcTable)
      .option("user", jdbcUser)
      .option("password", jdbcPassword)
      .option("driver", "oracle.jdbc.driver.OracleDriver")
      .load()

    val castedDF = downloadedDF.select(downloadedDF.columns.map(col => colcast(col, castedSchema.find(_.name == col).get.dataType)): _*)

    castedDF.write
      .format(hiveFormat)
      .mode("overwrite")
      .partitionBy(hivePartitionColumns:_*)
      .saveAsTable(s"$hiveDatabase.$hiveTable")

    //Change Data Capture of Oracle Table using Spark Streaming from JDBC
    val streamingDF = spark.readStream
      .format("jdbc")
      .option("url", jdbcUrl)
      .option("dbtable", jdbcTable)
      .option("user", jdbcUser)
      .option("password", jdbcPassword)
      .option("driver", "oracle.jdbc.driver.OracleDriver")
      .load()
      .schema(castedSchema)


    val streamingQuery = streamingDF.writeStream
      .format(hiveFormat)
      .option("checkpointLocation", hivePartitionLocation)
      .outputMode("append")
      .partitionBy(hivePartitionColumns:_*)
      .start(hiveDatabase + "." + hiveTable)

    streamingQuery.awaitTermination(streamingInterval * 1000)
  }
}

File: spark.py
#importing necessary libraries
from pyspark.sql import SparkSession
from pyspark.sql.types import *
from pyspark.sql.functions import *
from pyspark.sql.streaming import DataStreamWriter

#creating spark session
spark = SparkSession.builder.appName("OracleToHive").getOrCreate()

#creating oracle connection
oracle_url = "jdbc:oracle:thin:@localhost:1521:orcl"
oracle_properties = dict(user="username", password="password", driver="oracle.jdbc.driver.OracleDriver")

#creating hive connection
hive_url = "jdbc:hive2://localhost:10000/default"
hive_properties = dict(user="username", password="password", driver="org.apache.hive.jdbc.HiveDriver")

#creating oracle table
oracle_table = spark.read.jdbc(url=oracle_url, table="oracle_table", properties=oracle_properties)

#creating hive table
hive_table = spark.read.jdbc(url=hive_url, table="hive_table", properties=hive_properties)

#creating mapping for oracle data types
oracle_data_types_map = dict(NUMBER="DECIMAL", VARCHAR2="STRING", DATE="TIMESTAMP")

#creating schema for oracle table
oracle_schema = StructType([StructField(c.name, oracle_data_types_map.get(c.dataType, c.dataType), True) for c in oracle_table.schema.fields])

#downloading oracle table in hive with orc format
oracle_table.write.format("orc").mode("overwrite").saveAsTable("hive_table")

#creating streaming query
streaming_query = DataStreamWriter(oracle_table.writeStream.format("orc").option("checkpointLocation", "/tmp/checkpoint").start("/tmp/streaming"))

#appending streaming data to hive table
streaming_query.foreachBatch(lambda df, epoch_id: df.write.format("orc").mode("append").saveAsTable("hive_table"))

#stopping streaming query
streaming_query.stop()

#stopping spark session
spark.stop()
